hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -51,17 +69,46 @@ def url_encode(s):
 
 
 @st.defines_strategy_with_reusable_values
-def domains():
-    """A strategy for :rfc:`1035` fully qualified domain names."""
-    atoms = st.text(
-        string.ascii_letters + "0123456789-", min_size=1, max_size=63
-    ).filter(lambda s: "-" not in s[0] + s[-1])
+def domains(
+    max_length=255,  # type: int
+    max_element_length=63,  # type: int
+):
+    # type: (...) -> SearchStrategy[Text]
+    """A strategy for :rfc:`1035` fully qualified domain names.
+
+    The max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+    The maximum_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+    """
+    # https://tools.ietf.org/html/rfc1035#section-2.3.4
+    if max_length > 255:
+        raise InvalidArgument(
+            u"Invalid max_length %s. max_length must be shorter than 255.", max_length
+        )
+
+    if max_element_length > 63:
+        raise InvalidArgument(
+            u"Invalid max_element_length %s. max_element_length must be less than 63.",
+            max_element_length,
+        )
+
+    atoms = (
+        st.text(
+            string.ascii_letters + string.digits + "-",
+            min_size=1,
+            max_size=max_element_length,
+        )
+        .filter(lambda s: "-" not in s[0] + s[-1])
+        .filter(lambda s: bool(set(s) - set(string.digits)))
```

Comment:
What if the element is all digits and `-`? Is `-123` a valid element? Is `---`?

Also, continually rebuilding `set(string.digits)` is where this function spends almost half its time – if you cache it and add `"-"` you’ll be a tiny perf win! :D

Reply:
The filter lambda is there to catch starting and terminating `-` characters so would correctly reject both `-123` and `---`, but you've also made me realise that I forgot to add a filter lambda to keep out starting digits per the RFC, also nice catch on the repeatedly rebuilding the set, thats definitely a win, so I'll push an update with improvements here.

Reply:
After I shifted to pre-building the necessary combinations of character and sets of characters, simplify this to a two pass 'valid first' and 'valid last' which is a lot more readable, it does have some extra module level code comments to explain the pre-built variables, but I feel the trade off is worth it.